lib/HeroBonus.cpp

```diff
@@ -1253,14 +1310,37 @@ void CBonusSystemNode::getRedChildren(TNodes &out)
 	}
 }
 
-void CBonusSystemNode::newRedDescendant(CBonusSystemNode *descendant)
+void CBonusSystemNode::newRedDescendant(CBonusSystemNode * descendant)
 {
 	for(auto b : exportedBonuses)
+	{
 		if(b->propagator)
 			descendant->propagateBonus(b);
+	}
+	TNodes redParents;
+	getRedAncestors(redParents); //get all red parents recursively
 
-	FOREACH_RED_PARENT(parent)
-		parent->newRedDescendant(descendant);
+	//it's not in the 'getRedParents' due to infinite recursion loop.
+	//it's actual only for battle, otherwise, when garrison hero is linking to the town, town's 'children' is already empty
+	if(descendant->nodeType == BATTLE && nodeType == TOWN) //acts as a pure bonus source, but has bonus bearers.
+	{
+		for(auto child : children)
+		{
+			for(auto b : child->exportedBonuses)
+			{
+				if(b->propagator && b->propagator->getPropagatorType() == BATTLE)
+					descendant->propagateBonus(b);
+			}
+		}
+	}
+	for(auto parent : redParents)
+	{
+		for(auto b : parent->exportedBonuses)
+		{
+			if(b->propagator)
+				descendant->propagateBonus(b);
+		}
+	}

```

Comment:
```suggestion
	if(descendant->nodeType == BATTLE && nodeType == TOWN) //acts as a pure bonus source, but has bonus bearers.
	{
		for(auto child : children)
			for(auto b : child->exportedBonuses)
				if(b->propagator && b->propagator->getPropagatorType() == BATTLE)
					descendant->propagateBonus(b);
	}
	for(auto parent : redParents)
	{
		for(auto b : parent->exportedBonuses)
			if(b->propagator)
				descendant->propagateBonus(b);
	}

```

Reply:
![1](https://user-images.githubusercontent.com/58867270/116828248-28a8bd00-aba6-11eb-943d-0919c7f6bcd5.jpg)

Reply:
Changes have been merged to commit fa2649c70215ab81ac5f2c0a6627ae420514ef96